File: solutions/169.Majority_Element/AC_Moore_n.py
# Author:      illuz <iilluzen[at]gmail.com>
# File:        AC_Moore_n.py
# Create Date: 2015-02-05 10:48:49
# Descripton:   

class Solution:
    # @param num, a list of integers
    # @return an integer
    def majorityElement(self, num):
        cnt = 0
        for i in num:
            if cnt == 0:
                ret = i
            cnt += 1 if ret == i else -1
        return ret

# debug
s = Solution()
print s.majorityElement([1, 2, 2, 2, 3])

File: solutions/169.Majority_Element/AC_Moore_n.java
/*
*  Author:      illuz <iilluzen[at]gmail.com>
*  File:        AC_Moore_n.java
*  Create Date: 2015-02-05 10:48:42
*  Descripton:   
*/

import java.util.Scanner;

public class Solution {

    public int majorityElement(int[] num) {
        int cnt = 0;
        int major = 0;
        for (int i : num) {
            if (cnt == 0)
                major = i;
            cnt += (major == i) ? 1 : -1;
        }
        return major;
    }

    // debug
    public static void main(String[] args) {
        Scanner cin = new Scanner(System.in);
        Solution s = new Solution();
        System.out.println("no case");
    }

}
